# The Battle of Neighborhoods

## Toronto and New York City

# Introduction

Toronto is the capital city of the Canadian province of Ontario. With a recorded population of 2,731,571 in 2016, it is the most populous city in Canada and the fourth most populous city in North America. 

New York City is the most populous city in the United States. With an estimated 2019 population of 8,336,817 distributed over about 302.6 square miles, New York City is also the most densely populated major city in the United States.

Toronto and New York City are the popular vacation and migration destinations for people all around the world. They are diverse and multicultural and offer a wide variety of experiences that is widely sought after.  We try to group the neighborhoods of Toronto and New York City respectively and draw insights to what they look like now.

# Business Problem

The aim is to help people choose their destinations depending on the experiences that the neighborhoods have to offer and what they would want to have. This also helps people make decisions if they are thinking about migrating to Toronto or New York City or even if they want to relocate neighborhoods within the city. Our findings will help stakeholders make informed decisions and address any concerns they have including the different kinds of cuisines, provision stores and what the city has to offer.

# Data Description

We require geolocation data for both Toronto and New York City. Postal codes in each city serve as a starting point. Using Postal codes, we use can find out the neighborhoods, boroughs, venues and their most popular venue categories.

## Toronto

To derive our solution, we scrape our data from https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M

This Wikipedia page has information about all the neighborhoods, we limit it Toronto.

Also, it lacks of information about the geographical locations. To solve this problem, we use the Geocoder Python package.


## New York City

To derive our solution, we leverage JSON data available at https://geo.nyu.edu/catalog/nyu_2451_34572

## Foursquare API Data

We will need data about different venues in different neighborhoods of that specific borough. In order to gain that information, we will use "Foursquare" locational information. Foursquare is a location data provider with information about all manner of venues and events within an area of interest. Such information includes venue names, locations, menus and even photos. As such, the foursquare location platform will be used as the sole data source since all the stated required information can be obtained through the API.

After finding the list of neighborhoods, we then connect to the Foursquare API to gather information about venues inside every neighborhood. For each neighborhood, we have chosen the radius to be 500 meters.



The data retrieved from Foursquare contained information of venues within a specified distance of the longitude and latitude of the postcodes. The information obtained per venue as follows:

* *Neighborhood*: Name of the Neighborhood
* *Neighborhood Latitude*: Latitude of the Neighborhood
* *Neighborhood Longitude*: Longitude of the Neighborhood
* *Venue*: Name of the Venue
* *Venue Latitude*: Latitude of Venue
* *Venue Longitude*: Longitude of Venue
* *Venue Category*: Category of Venue


Based on all the information collected for both Toronto and New York City, we have sufficient data to build our model. We cluster the neighborhoods together based on similar venue categories. We then present our observations and findings. Using this data, our stakeholders can take the necessary decision.

# Methodology

We will be creating our model with the help of Python so we start off by importing all the required packages.

In [1]:
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup

import matplotlib.cm as cm
import matplotlib.colors as colors

!conda install -c conda-forge folium=0.5.0
import folium

from sklearn.cluster import KMeans

Solving environment: done

# All requested packages already installed.



Package breakdown:

* *Pandas* : To collect and manipulate data in JSON and HTMl and then data analysis
* *requests* : Handle http requests
* *matplotlib* : Detailing the generated maps
* *folium* : Generating maps of Toronto and New York City
* *sklearn* : To import Kmeans which is the machine learning model that we are using

The approach taken here is to explore each of the cities individually, plot the map to show the neighborhoods being considered and then build our model by clustering all the similar neighborhoods together and finally plot the new map with the clustered neighborhoods. We draw insights and then compare and discuss our findings.

# Exploring Toronto

## Neighborhoods of Toronto

We begin to start collecting and refining the data needed for the our business solution to work.

### Data Collection

To get the neighborhoods in Toronto, we start by scraping the list of postal codes of Canada Wikipedia page.

In [2]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
html_data  = requests.get(url).text

In [3]:
soup = BeautifulSoup(html_data,"html.parser")
soup.find_all('title')

[<title>List of postal codes of Canada: M - Wikipedia</title>]

In [4]:
table_contents = pd.DataFrame(columns=['Postal Code', 'Borough', 'Neighborhood'])

table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':  #To skip if the borough name is 'Not Assigned'
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)    

In [5]:
df=pd.DataFrame(table_contents)
df

PostalCode                                            Borough  \
0          M3A                                         North York   
1          M4A                                         North York   
2          M5A                                   Downtown Toronto   
3          M6A                                         North York   
4          M7A                                       Queen's Park   
..         ...                                                ...   
98         M8X                                          Etobicoke   
99         M4Y                                   Downtown Toronto   
100        M7Y  East TorontoBusiness reply mail Processing Cen...   
101        M8Y                                          Etobicoke   
102        M8Z                                          Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4                        Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100                                     Enclave of M4L  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

### Data Preprocessing

In [6]:
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})
df

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A           Queen's Park   
..         ...                    ...   
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y  East Toronto Business   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4                        Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100                                     Enclave of M4L  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

### Geolocations of the Toronto Neighborhoods

In [7]:
data = pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs_v1/Geospatial_Coordinates.csv")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

#### Since the dimensions of both dataframes are the same, we can join on postal code

#### Check the column types of both dataframes

In [8]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [9]:
df.dtypes

PostalCode      object
Borough         object
Neighborhood    object
dtype: object

#### Join the two dataframes

In [10]:
combined_data = df.join(data.set_index('Postal Code'), on='PostalCode', how='inner')
combined_data

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A           Queen's Park   
..         ...                    ...   
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y  East Toronto Business   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4                        Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100                                     Enclave of M4L  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

### Use geopy library to get the latitude and longitude values of Toronto

In [11]:
from geopy.geocoders import Nominatim
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


## Visualize the Map of Toronto

To help visualize the Map of Toronto and the neighborhoods in Toronto, we make use of the folium package.

In [12]:
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighborhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

## Explore Neighborhoods in Toronto

Using Foursquare API, we are able to get the venue and venue categories around each neighborhood in Toronto.

#### Define Foursquare Credentials and Version

In [13]:
CLIENT_ID = 'GDVLLEAOEYDPMLLND1PEIFP2UKLQ3TC42WEVXMSGCHVLT23S' 
CLIENT_SECRET = 'BV0P1UO5OLCHN0AOLTSAI2Q4SQQJEPKSC3TZK2E4GZXWNB2W'
VERSION = '20180605' # Foursquare API version
LIMIT = 100

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: GDVLLEAOEYDPMLLND1PEIFP2UKLQ3TC42WEVXMSGCHVLT23S
CLIENT_SECRET:BV0P1UO5OLCHN0AOLTSAI2Q4SQQJEPKSC3TZK2E4GZXWNB2W


#### Create a function to get all the venue categories in Toronto

In [14]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Collect the venues in Toronto for each neighborhood

In [15]:
venues_in_toronto = getNearbyVenues(combined_data['Neighborhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills North
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview East
The Danforth

In [16]:
venues_in_toronto.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0                     KFC       43.754387       -79.333021   
1         Brookbanks Park       43.751976       -79.332140   
2           Variety Store       43.751974       -79.333114   
3  Victoria Village Arena       43.723481       -79.315635   
4               Portugril       43.725819       -79.312785   

          Venue Category  
0   Fast Food Restaurant  
1                   Park  
2      Food & Drink Shop  
3           Hockey Arena  
4  Portuguese Restaurant

In [17]:
venues_in_toronto.shape

(2007, 7)

We have scraped total of 2007 records for venues.

#### Grouping by Neighborhoods

In [18]:
venues_in_toronto.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                             
Agincourt                                                            5   
Alderwood, Long Branch                                               8   
Bathurst Manor, Wilson Heights, Downsview North                     15   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   23   
...                                                                ...   
Willowdale West                                                      5   
Willowdale, Newtonbrook                                              1   
Woburn                                                               4   
Woodbine Heights                                                     4   
York Mills West                                                      2   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Agincourt                                                             5   
Alderwood, Long Branch                                                8   
Bathurst Manor, Wilson Heights, Downsview North                      15   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    23   
...                                                                 ...   
Willowdale West                                                       5   
Willowdale, Newtonbrook                                               1   
Woburn                                                                4   
Woodbine Heights                                                      4   
York Mills West                                                       2   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Agincourt                                            5               5   
Alderwood, Long Branch                               8               8   
Bathurst Manor, Wilson Heights, Downsview North     15              15   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   23              23   
...                                                ...             ...   
Willowdale West                                      5               5   
Willowdale, Newtonbrook                              1               1   
Woburn                                               4               4   
Woodbine Heights                                     4               4   
York Mills West                                      2               2   

                                                 Venue Longitude  \
Neighborhood                                                       
Agincourt                                                      5   
Alderwood, Long Branch                                         8   
Bathurst Manor, Wilson Heights, Downsview North               15   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             23   
...                                                          ...   
Willowdale West                                                5   
Willowdale, Newtonbrook                                        1   
Woburn                                                         4   
Woodbine Heights                                               4   
York Mills West                                                2   

                                                 Venue Category  
Neighborhood                                                     
Agincourt                                                     5  
Alderwood, Long Branch                                        8 

In [19]:
print('There are {} uniques categories.'.format(len(venues_in_toronto['Venue Category'].unique())))

There are 259 uniques categories.


### Analyze Each Neighborhood

#### One Hot Encoding

We need to Encode our venue categories to get a better result for our clustering.

In [20]:
# one hot encoding
Toronto_onehot = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Toronto_onehot['Neighborhood'] = venues_in_toronto['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head()

Yoga Studio  Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0            0                  0               0                  0        0   
1            0                  0               0                  0        0   
2            0                  0               0                  0        0   
3            0                  0               0                  0        0   
4            0                  0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  ...  Truck Stop  Vegetarian / Vegan Restaurant  \
0                 0  ...           0                              0   
1                 0  ...           0                              0   
2                 0  ...           0                              0   
3                 0  ...           0                              0   
4                 0  ...           0                              0   

   Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                 0            0                      0                0   
1                 0            0                      0                0   
2                 0            0                      0                0   
3                 0            0                      0                0   
4                 0            0                      0                0   

   Wine Bar  Wine Shop  Wings Joint  Women's Store  
0         0          0            0              0  
1         0          0            0              0  
2         0          0            0              0  
3         0          0            0              0  
4         0          0            0              0  

[5 rows x 259 columns]

### Venue categories mean value

We will group the neighborhoods and calculate the mean of the frequency of occurrence of each category.

In [21]:
Toronto_grouped = Toronto_onehot.groupby('Neighborhood').mean().reset_index()
Toronto_grouped

Neighborhood  Yoga Studio  \
0                                         Agincourt          0.0   
1                            Alderwood, Long Branch          0.0   
2   Bathurst Manor, Wilson Heights, Downsview North          0.0   
3                                   Bayview Village          0.0   
4                 Bedford Park, Lawrence Manor East          0.0   
..                                              ...          ...   
95                                  Willowdale West          0.0   
96                          Willowdale, Newtonbrook          0.0   
97                                           Woburn          0.0   
98                                 Woodbine Heights          0.0   
99                                  York Mills West          0.0   

    Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0                 0.0             0.0                0.0      0.0   
1                 0.0             0.0                0.0      0.0   
2                 0.0             0.0                0.0      0.0   
3                 0.0             0.0                0.0      0.0   
4                 0.0             0.0                0.0      0.0   
..                ...             ...                ...      ...   
95                0.0             0.0                0.0      0.0   
96                0.0             0.0                0.0      0.0   
97                0.0             0.0                0.0      0.0   
98                0.0             0.0                0.0      0.0   
99                0.0             0.0                0.0      0.0   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  ...  \
0                  0.0           0.0             0.0              0.0  ...   
1                  0.0           0.0             0.0              0.0  ...   
2                  0.0           0.0             0.0              0.0  ...   
3                  0.0           0.0             0.0              0.0  ...   
4                  0.0           0.0             0.0              0.0  ...   
..                 ...           ...             ...              ...  ...   
95                 0.0           0.0             0.0              0.0  ...   
96                 0.0           0.0             0.0              0.0  ...   
97                 0.0           0.0             0.0              0.0  ...   
98                 0.0           0.0             0.0              0.0  ...   
99                 0.0           0.0             0.0              0.0  ...   

    Truck Stop  Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0          0.0                            0.0               0.0         0.00   
1          0.0                            0.0               0.0         0.00   
2          0.0                            0.0               0.0         0.00   
3          0.0                            0.0               0.0         0.00   
4          0.0                            0.0               0.0         0.00   
..         ...                            ...               ...          ...   
95         0.0                            0.0               0.0         0.00   
96         0.0                            0.0               0.0         0.00   
97         0.0                            0.0               0.0         0.00   
98         0.0                            0.0               0.0         0.25   
99         0.0                            0.0               0.0         0.00   

    Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                     0.0              0.0       0.0        0.0          0.0   
1                     0.0              0.0       0.0        0.0          0.0   
2                     0.0              0.0       0.0        0.0          0.0   
3                     0.0              0.0       0.0        0.0          0.0   
4                     0.0              0.0       0.0        0.0          0.0   
..             

#### Create a function to get the top most common venue categories

In [22]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

There are too many venue categories, we can take the top 10 to cluster the neighborhoods.

Creating a function to label the columns of the venue correctly.

In [23]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

### Top venue categories

Getting the top venue categories in Toronto

In [24]:
# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Toronto_grouped['Neighborhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt        Breakfast Spot   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village   Japanese Restaurant   
4                Bedford Park, Lawrence Manor East            Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Lounge          Skating Rink        Clothing Store   
1                   Pub        Sandwich Place            Playground   
2           Coffee Shop         Shopping Mall      Sushi Restaurant   
3                  Café                  Bank    Chinese Restaurant   
4    Italian Restaurant           Coffee Shop        Sandwich Place   

       5th Most Common Venue      6th Most Common Venue  \
0  Latin American Restaurant  Middle Eastern Restaurant   
1               Dance Studio                Coffee Shop   
2                      Diner                Gas Station   
3                     Museum                      Motel   
4          Indian Restaurant           Sushi Restaurant   

             7th Most Common Venue            8th Most Common Venue  \
0  Molecular Gastronomy Restaurant       Modern European Restaurant   
1                     Skating Rink  Molecular Gastronomy Restaurant   
2                   Sandwich Place               Frozen Yogurt Shop   
3              Moroccan Restaurant              Monument / Landmark   
4                             Café                     Liquor Store   

             9th Most Common Venue      10th Most Common Venue  
0                Mobile Phone Shop          Miscellaneous Shop  
1       Modern European Restaurant           Mobile Phone Shop  
2                   Ice Cream Shop                 Bridal Shop  
3  Molecular Gastronomy Restaurant  Modern European Restaurant  
4                          Butcher             Thai Restaurant

## Model Building

### K-Means

Let's cluster the city of Toronto to roughly 5 to make it easier to analyze.

We use the K Means clustering technique to do so.

In [25]:
# Run k-means to cluster the neighborhood into 5 clusters

# set number of clusters
kclusters = 5

Toronto_grouped_clustering = Toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 4], dtype=int32)

### Labelling Clustered Data

In [26]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Toronto_merged = combined_data

# merge Toronto_grouped with Toronto_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Toronto_merged.head()

PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  Cluster Labels 1st Most Common Venue   2nd Most Common Venue  \
0 -79.329656             4.0  Fast Food Restaurant                    Park   
1 -79.315572             0.0            Nail Salon           Grocery Store   
2 -79.360636             0.0           Coffee Shop                    Park   
3 -79.464763             0.0        Clothing Store  Furniture / Home Store   
4 -79.389494             0.0           Coffee Shop        Sushi Restaurant   

  3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
0     Food & Drink Shop   Martial Arts School       Malay Restaurant   
1           Coffee Shop          Hockey Arena  Portuguese Restaurant   
2                   Pub                Bakery                   Café   
3    Miscellaneous Shop     Accessories Store               Boutique   
4         Burrito Place           Yoga Studio            Salad Place   

   6th Most Common Venue     7th Most Common Venue 8th Most Common Venue  \
0         Medical Center  Mediterranean Restaurant           Men's Store   
1            Yoga Studio        Miscellaneous Shop   Monument / Landmark   
2             Restaurant            Farmers Market      Sushi Restaurant   
3  Vietnamese Restaurant               Coffee Shop   Moroccan Restaurant   
4     Mexican Restaurant              Burger Joint                  Café   

             9th Most Common Venue           10th Most Common Venue  
0              Moroccan Restaurant                    Metro Station  
1  Molecular Gastronomy Restaurant       Modern European Restaurant  
2                       Beer Store            Performing Arts Venue  
3              Monument / Landmark  Molecular Gastronomy Restaurant  
4                   Sandwich Place                       Restaurant

#### Drop all the NaN values to prevent data skew

In [27]:
Toronto_merged_nonan = Toronto_merged.dropna(subset=['Cluster Labels'])

### Visualizing the Clustered neighborhood

In [28]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged_nonan['Latitude'], Toronto_merged_nonan['Longitude'], Toronto_merged_nonan['Neighborhood'], Toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

## Examining our Clusters

#### Cluster 1

In [29]:
Toronto_merged_nonan.loc[Toronto_merged_nonan['Cluster Labels'] == 0, Toronto_merged_nonan.columns[[1] + list(range(5, Toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1               North York             0.0            Nail Salon   
2         Downtown Toronto             0.0           Coffee Shop   
3               North York             0.0        Clothing Store   
4             Queen's Park             0.0           Coffee Shop   
7               North York             0.0   Japanese Restaurant   
..                     ...             ...                   ...   
97        Downtown Toronto             0.0           Coffee Shop   
98               Etobicoke             0.0                 River   
99        Downtown Toronto             0.0      Sushi Restaurant   
100  East Toronto Business             0.0           Yoga Studio   
102              Etobicoke             0.0         Burrito Place   

      2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
1             Grocery Store           Coffee Shop               Hockey Arena   
2                      Park                   Pub                     Bakery   
3    Furniture / Home Store    Miscellaneous Shop          Accessories Store   
4          Sushi Restaurant         Burrito Place                Yoga Studio   
7            Baseball Field  Caribbean Restaurant                        Gym   
..                      ...                   ...                        ...   
97                     Café        Sandwich Place                      Hotel   
98                     Pool           Yoga Studio  Middle Eastern Restaurant   
99      Japanese Restaurant           Coffee Shop                    Gay Bar   
100                 Brewery                Garden         Light Rail Station   
102          Sandwich Place           Wings Joint              Tanning Salon   

     5th Most Common Venue            6th Most Common Venue  \
1    Portuguese Restaurant                      Yoga Studio   
2                     Café                       Restaurant   
3                 Boutique            Vietnamese Restaurant   
4              Salad Place               Mexican Restaurant   
7                     Café                    Movie Theater   
..                     ...                              ...   
97     Japanese Restaurant                             Bank   
98     Monument / Landmark  Molecular Gastronomy Restaurant   
99              Restaurant                Indian Restaurant   
100                    Spa             Fast Food Restaurant   
102      Convenience Store                      Flower Shop   

          7th Most Common Venue 8th Most Common Venue  \
1            Miscellaneous Shop   Monument / Landmark   
2                Farmers Market      Sushi Restaurant   
3                   Coffee Shop   Moroccan Restaurant   
4                  Burger Joint                  Café   
7                         Motel   Moroccan Restaurant   
..                          ...                   ...   
97                          Gym      Asian Restaurant   
98   Modern European Restaurant     Mobile Phone Shop   
99                Burrito Place                   Gym   
100                  Smoke Shop        Farmers Market   
102              Discount Store                   Gym   

               9th Most Common Venue           10th Most Common Venue  
1    Molecular Gastronomy Restaurant       Modern European Restaurant  
2                         Beer Store            Performing Arts Venue  
3                Monument / Landmark  Molecular Gastronomy Restaurant  
4                     Sandwich Place                       Restaurant  
7                Monument / Landmark  Molecular Gastronomy Restaurant  
..                               ...                              ...  
97                        Restaurant                 Sushi Restaurant  
98                Miscellaneous Shop               Mexican Restaurant  
99              Fast Food Restaurant         Mediterranean Restaurant  
100                       Skate Park                    Auto Workshop  
102    

#### Cluster 2

In [30]:
Toronto_merged_nonan.loc[Toronto_merged_nonan['Cluster Labels'] == 1, Toronto_merged_nonan.columns[[1] + list(range(5, Toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
52  North York             1.0                  Park           Yoga Studio   
64        York             1.0                  Park           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
52   Moroccan Restaurant      Malay Restaurant   Martial Arts School   
64   Moroccan Restaurant      Malay Restaurant   Martial Arts School   

   6th Most Common Venue     7th Most Common Venue 8th Most Common Venue  \
52        Medical Center  Mediterranean Restaurant           Men's Store   
64        Medical Center  Mediterranean Restaurant           Men's Store   

   9th Most Common Venue 10th Most Common Venue  
52         Metro Station     Mexican Restaurant  
64         Metro Station     Mexican Restaurant

#### Cluster 3

In [31]:
Toronto_merged_nonan.loc[Toronto_merged_nonan['Cluster Labels'] == 2, Toronto_merged_nonan.columns[[1] + list(range(5, Toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
57   North York             2.0        Baseball Field          Food Service   
101   Etobicoke             2.0        Baseball Field           Yoga Studio   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
57            Yoga Studio   Moroccan Restaurant   Martial Arts School   
101                 Motel   Martial Arts School        Medical Center   

        6th Most Common Venue     7th Most Common Venue 8th Most Common Venue  \
57             Medical Center  Mediterranean Restaurant           Men's Store   
101  Mediterranean Restaurant               Men's Store         Metro Station   

    9th Most Common Venue     10th Most Common Venue  
57          Metro Station         Mexican Restaurant  
101    Mexican Restaurant  Middle Eastern Restaurant

#### Cluster 4

In [32]:
Toronto_merged_nonan.loc[Toronto_merged_nonan['Cluster Labels'] == 3, Toronto_merged_nonan.columns[[1] + list(range(5, Toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough             3.0  Fast Food Restaurant    Mac & Cheese Joint   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6      Malay Restaurant   Martial Arts School        Medical Center   

      6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6  Mediterranean Restaurant           Men's Store         Metro Station   

  9th Most Common Venue     10th Most Common Venue  
6    Mexican Restaurant  Middle Eastern Restaurant

#### Cluster 5

In [33]:
Toronto_merged_nonan.loc[Toronto_merged_nonan['Cluster Labels'] == 4, Toronto_merged_nonan.columns[[1] + list(range(5, Toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0               North York             4.0  Fast Food Restaurant   
21                    York             4.0                  Park   
35  East York/East Toronto             4.0     Convenience Store   
66              North York             4.0     Convenience Store   
77               Etobicoke             4.0                  Park   
83         Central Toronto             4.0          Tennis Court   
85             Scarborough             4.0          Intersection   
91        Downtown Toronto             4.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
0                   Park     Food & Drink Shop        Martial Arts School   
21         Women's Store                  Pool  Middle Eastern Restaurant   
35                  Park    Mexican Restaurant        Monument / Landmark   
66                  Park    Mexican Restaurant        Monument / Landmark   
77     Mobile Phone Shop        Sandwich Place                Yoga Studio   
83                  Park            Playground                      Trail   
85            Playground                  Park  Middle Eastern Restaurant   
91            Playground                 Trail                Yoga Studio   

              5th Most Common Venue            6th Most Common Venue  \
0                  Malay Restaurant                   Medical Center   
21              Monument / Landmark  Molecular Gastronomy Restaurant   
35  Molecular Gastronomy Restaurant       Modern European Restaurant   
66  Molecular Gastronomy Restaurant       Modern European Restaurant   
77              Moroccan Restaurant              Martial Arts School   
83        Middle Eastern Restaurant              Monument / Landmark   
85              Moroccan Restaurant              Monument / Landmark   
91               Mexican Restaurant  Molecular Gastronomy Restaurant   

              7th Most Common Venue       8th Most Common Venue  \
0          Mediterranean Restaurant                 Men's Store   
21       Modern European Restaurant           Mobile Phone Shop   
35                Mobile Phone Shop          Miscellaneous Shop   
66                Mobile Phone Shop          Miscellaneous Shop   
77                   Medical Center    Mediterranean Restaurant   
83  Molecular Gastronomy Restaurant  Modern European Restaurant   
85  Molecular Gastronomy Restaurant  Modern European Restaurant   
91       Modern European Restaurant           Mobile Phone Shop   

        9th Most Common Venue     10th Most Common Venue  
0         Moroccan Restaurant              Metro Station  
21         Miscellaneous Shop              Metro Station  
35  Middle Eastern Restaurant                Yoga Studio  
66  Middle Eastern Restaurant                Yoga Studio  
77                Men's Store              Metro Station  
83          Mobile Phone Shop         Miscellaneous Shop  
85          Mobile Phone Shop         Miscellaneous Shop  
91         Miscellaneous Shop  Middle Eastern Restaurant

# Exploring New York City

## Neighborhoods of New York City

### Data Collection

We load the json data.

In [34]:
!wget -q -O 'newyork_data.json' https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs/newyork_data.json
print('Data downloaded!')

Data downloaded!


In [36]:
import json

with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

In [37]:
newyork_data

{'type': 'FeatureCollection',
 'totalFeatures': 306,
 'features': [{'type': 'Feature',
   'id': 'nyu_2451_34572.1',
   'geometry': {'type': 'Point',
    'coordinates': [-73.84720052054902, 40.89470517661]},
   'geometry_name': 'geom',
   'properties': {'name': 'Wakefield',
    'stacked': 1,
    'annoline1': 'Wakefield',
    'annoline2': None,
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.84720052054902,
     40.89470517661,
     -73.84720052054902,
     40.89470517661]}},
  {'type': 'Feature',
   'id': 'nyu_2451_34572.2',
   'geometry': {'type': 'Point',
    'coordinates': [-73.82993910812398, 40.87429419303012]},
   'geometry_name': 'geom',
   'properties': {'name': 'Co-op City',
    'stacked': 2,
    'annoline1': 'Co-op',
    'annoline2': 'City',
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.82993910812398,
     40.87429419303012,
     -73.82993910812398,
     40.87429419303012]}},
  {'type': 'Feature',
 

### Feature Selection

In [38]:
neighborhoods_data = newyork_data['features']

In [39]:
neighborhoods_data[0]

{'type': 'Feature',
 'id': 'nyu_2451_34572.1',
 'geometry': {'type': 'Point',
  'coordinates': [-73.84720052054902, 40.89470517661]},
 'geometry_name': 'geom',
 'properties': {'name': 'Wakefield',
  'stacked': 1,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'annoangle': 0.0,
  'borough': 'Bronx',
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661]}}

#### Tranform the data into a pandas dataframe.

In [40]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)

for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)
    
neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

### Geolocations of the New York City Neighborhoods

#### Use geopy library to get the latitude and longitude values of New York City.

In [41]:
address = 'New York City, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 40.7127281, -74.0060152.


## Visualize the Map of New York City

In [42]:
# create map of New York using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

For illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods in Manhattan. So let's slice the original dataframe and create a new dataframe of the Manhattan data.

In [43]:
manhattan_data = neighborhoods[neighborhoods['Borough'] == 'Manhattan'].reset_index(drop=True)
manhattan_data.head()

Borough        Neighborhood   Latitude  Longitude
0  Manhattan         Marble Hill  40.876551 -73.910660
1  Manhattan           Chinatown  40.715618 -73.994279
2  Manhattan  Washington Heights  40.851903 -73.936900
3  Manhattan              Inwood  40.867684 -73.921210
4  Manhattan    Hamilton Heights  40.823604 -73.949688

Let's get the geographical coordinates of Manhattan.

In [44]:
address = 'Manhattan, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Manhattan are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Manhattan are 40.7896239, -73.9598939.


## Visualize the Map of Manhattan

In [45]:
# create map of Manhattan using latitude and longitude values
map_manhattan = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(manhattan_data['Latitude'], manhattan_data['Longitude'], manhattan_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_manhattan)  
    
map_manhattan

# Explore Neighborhoods in Manhattan

Using Foursquare API, we are able to get the venue and venue categories around each neighborhood in Manhattan.

### Venues in Manhattan

Using our previously defined function. Let's get the neaby venues present in each neighborhood of Manhattan.

In [46]:
manhattan_venues = getNearbyVenues(names=manhattan_data['Neighborhood'],latitudes=manhattan_data['Latitude'], longitudes=manhattan_data['Longitude'])

Marble Hill
Chinatown
Washington Heights
Inwood
Hamilton Heights
Manhattanville
Central Harlem
East Harlem
Upper East Side
Yorkville
Lenox Hill
Roosevelt Island
Upper West Side
Lincoln Square
Clinton
Midtown
Murray Hill
Chelsea
Greenwich Village
East Village
Lower East Side
Tribeca
Little Italy
Soho
West Village
Manhattan Valley
Morningside Heights
Gramercy
Battery Park City
Financial District
Carnegie Hill
Noho
Civic Center
Midtown South
Sutton Place
Turtle Bay
Tudor City
Stuyvesant Town
Flatiron
Hudson Yards


In [47]:
manhattan_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude          Venue  \
0  Marble Hill              40.876551               -73.91066    Bikram Yoga   
1  Marble Hill              40.876551               -73.91066       Arturo's   
2  Marble Hill              40.876551               -73.91066  Tibbett Diner   
3  Marble Hill              40.876551               -73.91066       Rite Aid   
4  Marble Hill              40.876551               -73.91066         Subway   

   Venue Latitude  Venue Longitude  Venue Category  
0       40.876844       -73.906204     Yoga Studio  
1       40.874412       -73.910271     Pizza Place  
2       40.880404       -73.908937           Diner  
3       40.875467       -73.908906        Pharmacy  
4       40.874667       -73.909586  Sandwich Place

In [49]:
manhattan_venues.shape

(2996, 7)

We have managed to collect 2996 venue records for the neighborhoods in Manhattan.

### Grouping by Neighborhoods

In [50]:
manhattan_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                                
Battery Park City                       62                      62     62   
Carnegie Hill                           71                      71     71   
Central Harlem                          37                      37     37   
Chelsea                                100                     100    100   
Chinatown                              100                     100    100   
Civic Center                            86                      86     86   
Clinton                                100                     100    100   
East Harlem                             41                      41     41   
East Village                           100                     100    100   
Financial District                     100                     100    100   
Flatiron                               100                     100    100   
Gramercy                                61                      61     61   
Greenwich Village                      100                     100    100   
Hamilton Heights                        55                      55     55   
Hudson Yards                            58                      58     58   
Inwood                                  53                      53     53   
Lenox Hill                             100                     100    100   
Lincoln Square                          92                      92     92   
Little Italy                           100                     100    100   
Lower East Side                         37                      37     37   
Manhattan Valley                        38                      38     38   
Manhattanville                          40                      40     40   
Marble Hill                             24                      24     24   
Midtown                                100                     100    100   
Midtown South                          100                     100    100   
Morningside Heights                     40                      40     40   
Murray Hill                             97                      97     97   
Noho                                   100                     100    100   
Roosevelt Island                        28                      28     28   
Soho                                   100                     100    100   
Stuyvesant Town                         18                      18     18   
Sutton Place                            90                      90     90   
Tribeca                                 78                      78     78   
Tudor City                              74                      74     74   
Turtle Bay                              92                      92     92   
Upper East Side                         76                      76     76   
Upper West Side                         69                      69     69   
Washington Heights                      82                      82     82   
West Village                           100                     100    100   
Yorkville                               97                      97     97   

                     Venue Latitude  Venue Longitude  Venue Category  
Neighborhood                                                          
Battery Park City                62               62              62  
Carnegie Hill                    71               71              71  
Central Harlem                   37               37              37  
Chelsea                         100              100             100  
Chinatown                       100              100             100  
Civic Center                     86               86              86  
Clinton                         100              100             100  
East Harlem                      41               41              41  
East Village                    100              100             100  
Finan

In [51]:
print('There are {} uniques categories.'.format(len(manhattan_venues['Venue Category'].unique())))

There are 326 uniques categories.


### Analyze Each Neighborhood

#### One Hot Encoding

We need to Encode our venue categories to get a better result for our clustering.

In [52]:
# one hot encoding
manhattan_onehot = pd.get_dummies(manhattan_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
manhattan_onehot['Neighborhood'] = manhattan_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [manhattan_onehot.columns[-1]] + list(manhattan_onehot.columns[:-1])
manhattan_onehot = manhattan_onehot[fixed_columns]

manhattan_onehot.head()

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0  Marble Hill                  0               0                  0   
1  Marble Hill                  0               0                  0   
2  Marble Hill                  0               0                  0   
3  Marble Hill                  0               0                  0   
4  Marble Hill                  0               0                  0   

   African Restaurant  American Restaurant  Antique Shop  \
0                   0                    0             0   
1                   0                    0             0   
2                   0                    0             0   
3                   0                    0             0   
4                   0                    0             0   

   Argentinian Restaurant  Art Gallery  Art Museum  ...  \
0                       0            0           0  ...   
1                       0            0           0  ...   
2                       0            0           0  ...   
3                       0            0           0  ...   
4                       0            0           0  ...   

   Vietnamese Restaurant  Volleyball Court  Waterfront  Weight Loss Center  \
0                      0                 0           0                   0   
1                      0                 0           0                   0   
2                      0                 0           0                   0   
3                      0                 0           0                   0   
4                      0                 0           0                   0   

   Whisky Bar  Wine Bar  Wine Shop  Wings Joint  Women's Store  Yoga Studio  
0           0         0          0            0              0            1  
1           0         0          0            0              0            0  
2           0         0          0            0              0            0  
3           0         0          0            0              0            0  
4           0         0          0            0              0            0  

[5 rows x 327 columns]

### Venue categories mean value

We will group the neighborhoods and calculate the mean of the frequency of occurrence of each category.

In [53]:
manhattan_grouped = manhattan_onehot.groupby('Neighborhood').mean().reset_index()
manhattan_grouped

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0     Battery Park City           0.000000        0.000000               0.00   
1         Carnegie Hill           0.000000        0.000000               0.00   
2        Central Harlem           0.000000        0.000000               0.00   
3               Chelsea           0.000000        0.000000               0.00   
4             Chinatown           0.000000        0.000000               0.00   
5          Civic Center           0.000000        0.000000               0.00   
6               Clinton           0.000000        0.000000               0.00   
7           East Harlem           0.000000        0.000000               0.00   
8          East Village           0.000000        0.000000               0.00   
9    Financial District           0.010000        0.000000               0.00   
10             Flatiron           0.000000        0.000000               0.00   
11             Gramercy           0.000000        0.000000               0.00   
12    Greenwich Village           0.010000        0.000000               0.00   
13     Hamilton Heights           0.000000        0.018182               0.00   
14         Hudson Yards           0.000000        0.000000               0.00   
15               Inwood           0.000000        0.000000               0.00   
16           Lenox Hill           0.000000        0.000000               0.01   
17       Lincoln Square           0.010870        0.000000               0.00   
18         Little Italy           0.000000        0.000000               0.00   
19      Lower East Side           0.000000        0.000000               0.00   
20     Manhattan Valley           0.000000        0.000000               0.00   
21       Manhattanville           0.000000        0.000000               0.00   
22          Marble Hill           0.000000        0.000000               0.00   
23              Midtown           0.000000        0.000000               0.00   
24        Midtown South           0.000000        0.000000               0.00   
25  Morningside Heights           0.000000        0.000000               0.00   
26          Murray Hill           0.000000        0.000000               0.00   
27                 Noho           0.000000        0.000000               0.00   
28     Roosevelt Island           0.000000        0.000000               0.00   
29                 Soho           0.000000        0.000000               0.00   
30      Stuyvesant Town           0.000000        0.000000               0.00   
31         Sutton Place           0.000000        0.011111               0.00   
32              Tribeca           0.000000        0.000000               0.00   
33           Tudor City           0.000000        0.000000               0.00   
34           Turtle Bay           0.000000        0.000000               0.00   
35      Upper East Side           0.000000        0.000000               0.00   
36      Upper West Side           0.000000        0.000000               0.00   
37   Washington Heights           0.012195        0.000000               0.00   
38         West Village           0.010000        0.010000               0.00   
39            Yorkville           0.000000        0.000000               0.00   

    African Restaurant  American Restaurant  Antique Shop  \
0             0.000000             0.000000      0.000000   
1             0.000000             0.014085      0.000000   
2             0.081081             0.054054      0.000000   
3             0.000000             0.030000      0.000000   
4             0.000000             0.030000      0.000000   
5             0.000000             0.011628      0.011628   
6             0.000000             0.050000      0.000000   
7             0.000000             0.000000      0.000000   
8             0.000000             0.010000      0.000000   
9             0.000000             0.030000      0.000000   
10            0.00

### Top venue categories

Reusing our previously defined function to get the top venue categories in the neighborhoods of Manhattan.

In [54]:
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = manhattan_grouped['Neighborhood']

for ind in np.arange(manhattan_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(manhattan_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0  Battery Park City                  Park                 Hotel   
1      Carnegie Hill           Coffee Shop             Wine Shop   
2     Central Harlem    African Restaurant        Sandwich Place   
3            Chelsea           Art Gallery           Coffee Shop   
4          Chinatown                Bakery    Chinese Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Coffee Shop                   Gym         Memorial Site   
1           Pizza Place        Cosmetics Shop                  Café   
2  Gym / Fitness Center     French Restaurant                   Bar   
3                Bakery     French Restaurant   American Restaurant   
4          Cocktail Bar   American Restaurant                   Spa   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0            Food Court                 Plaza         Boat or Ferry   
1                   Gym           Yoga Studio                   Pub   
2    Seafood Restaurant   Fried Chicken Joint   American Restaurant   
3             Wine Shop    Italian Restaurant        Ice Cream Shop   
4          Dessert Shop    Salon / Barbershop        Ice Cream Shop   

  9th Most Common Venue 10th Most Common Venue  
0            Playground            Pizza Place  
1  Gym / Fitness Center       Sushi Restaurant  
2                  Bank                Library  
3      Tapas Restaurant                  Hotel  
4          Optical Shop           Noodle House

## Model Building

### K Means

Let's cluster the neighborhoods to roughly 5 to make it easier to analyze.

We use the K Means clustering technique to do so.

In [55]:
# set number of clusters
kclusters = 5

manhattan_grouped_clustering = manhattan_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(manhattan_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 1, 2, 1, 1, 1, 1, 0, 1, 1], dtype=int32)

### Labelling Clustered Data

In [56]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

manhattan_merged = manhattan_data

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
manhattan_merged = manhattan_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

manhattan_merged.head()

Borough        Neighborhood   Latitude  Longitude  Cluster Labels  \
0  Manhattan         Marble Hill  40.876551 -73.910660               4   
1  Manhattan           Chinatown  40.715618 -73.994279               1   
2  Manhattan  Washington Heights  40.851903 -73.936900               2   
3  Manhattan              Inwood  40.867684 -73.921210               2   
4  Manhattan    Hamilton Heights  40.823604 -73.949688               0   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0        Sandwich Place      Department Store       Supplement Shop   
1                Bakery    Chinese Restaurant          Cocktail Bar   
2                  Café           Pizza Place                Bakery   
3    Mexican Restaurant                  Café                Lounge   
4           Pizza Place                  Café        Sandwich Place   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0      Storage Facility            Steakhouse        Clothing Store   
1   American Restaurant                   Spa          Dessert Shop   
2    Chinese Restaurant     Mobile Phone Shop         Grocery Store   
3                  Bank           Pizza Place                  Park   
4    Mexican Restaurant                Bakery         Deli / Bodega   

  7th Most Common Venue 8th Most Common Venue    9th Most Common Venue  \
0           Coffee Shop    Seafood Restaurant            Deli / Bodega   
1    Salon / Barbershop        Ice Cream Shop             Optical Shop   
2                  Bank                  Park  New American Restaurant   
3         Deli / Bodega                Bakery                 Wine Bar   
4           Coffee Shop          Cocktail Bar     Caribbean Restaurant   

  10th Most Common Venue  
0            Yoga Studio  
1           Noodle House  
2        Supplement Shop  
3   Caribbean Restaurant  
4      Indian Restaurant

### Visualizing the Clustered neighborhood

In [57]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(manhattan_merged['Latitude'], manhattan_merged['Longitude'], manhattan_merged['Neighborhood'], manhattan_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examining our Clusters

#### Cluster 1

In [58]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 0, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
4   Hamilton Heights           Pizza Place                  Café   
7        East Harlem    Mexican Restaurant                Bakery   
20   Lower East Side                  Café           Art Gallery   
25  Manhattan Valley           Yoga Studio    Mexican Restaurant   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
4         Sandwich Place    Mexican Restaurant                Bakery   
7            Pizza Place            Steakhouse                  Bank   
20                Bakery    Chinese Restaurant           Pizza Place   
25           Pizza Place           Coffee Shop     Indian Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
4          Deli / Bodega           Coffee Shop          Cocktail Bar   
7         Sandwich Place           Gas Station    Spanish Restaurant   
20        Sandwich Place                 Diner              Pet Café   
25                  Café                Hostel       Bubble Tea Shop   

   9th Most Common Venue   10th Most Common Venue  
4   Caribbean Restaurant        Indian Restaurant  
7                    Spa  New American Restaurant  
20           Tailor Shop      Filipino Restaurant  
25            Playground             Liquor Store

#### Cluster 2

In [59]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 1, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
1            Chinatown                Bakery    Chinese Restaurant   
8      Upper East Side    Italian Restaurant                   Spa   
9            Yorkville    Italian Restaurant      Sushi Restaurant   
10          Lenox Hill    Italian Restaurant                  Café   
12     Upper West Side                Bakery    Italian Restaurant   
14             Clinton               Theater    Italian Restaurant   
15             Midtown                 Hotel        Clothing Store   
16         Murray Hill           Coffee Shop   Japanese Restaurant   
17             Chelsea           Art Gallery           Coffee Shop   
18   Greenwich Village    Italian Restaurant        Clothing Store   
19        East Village                   Bar              Wine Bar   
21             Tribeca                  Park                   Spa   
22        Little Italy                Bakery    Italian Restaurant   
23                Soho        Clothing Store    Italian Restaurant   
24        West Village    Italian Restaurant          Cocktail Bar   
27            Gramercy           Pizza Place    Italian Restaurant   
29  Financial District           Coffee Shop           Pizza Place   
30       Carnegie Hill           Coffee Shop             Wine Shop   
31                Noho           Coffee Shop    Italian Restaurant   
32        Civic Center           Coffee Shop                   Spa   
33       Midtown South     Korean Restaurant                 Hotel   
34        Sutton Place    Italian Restaurant                  Park   
38            Flatiron    Italian Restaurant   American Restaurant   

            3rd Most Common Venue    4th Most Common Venue  \
1                    Cocktail Bar      American Restaurant   
8                         Exhibit                   Bakery   
9             Japanese Restaurant              Coffee Shop   
10                    Pizza Place                     Bank   
12                           Café                 Wine Bar   
14                    Coffee Shop      American Restaurant   
15                        Theater                      Spa   
16                          Hotel     Gym / Fitness Center   
17                         Bakery        French Restaurant   
18                   Dessert Shop         Sushi Restaurant   
19  Vegetarian / Vegan Restaurant           Ice Cream Shop   
21             Italian Restaurant                     Café   
22                           Café                    Hotel   
23                     Shoe Store                   Bakery   
24        New American Restaurant                 Wine Bar   
27                    Coffee Shop           Sandwich Place   
29                   Cocktail Bar     Gym / Fitness Center   
30                    Pizza Place           Cosmetics Shop   
31                   Cocktail Bar            Grocery Store   
32              French Restaurant     Gym / Fitness Center   
33                      Hotel Bar                   Bakery   
34                     Steakhouse                      Spa   
38                            Spa  New American Restaurant   

   5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
1                    Spa          Dessert Shop         Salon / Barbershop   
8              Juice Bar           Coffee Shop                      Hotel   
9                    Gym                   Bar                      Diner   
10      Sushi Restaurant                   Gym               Burger Joint   
12           Coffee Shop        Cosmetics Shop  Middle Eastern Restaurant   
14        Sandwich Place  Gym / Fitness Center                        Gym   
15            Food Truck      Sushi Restaurant                Coffee Shop   
16                   Spa           Pizza Place        American Restaurant   
17   American Restaurant             Wine Shop         Italian Restaurant   
18                  Café     Indian Restaurant                Coffee S

#### Cluster 3

In [60]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 2, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
2    Washington Heights                  Café           Pizza Place   
3                Inwood    Mexican Restaurant                  Café   
5        Manhattanville           Coffee Shop                  Bank   
6        Central Harlem    African Restaurant        Sandwich Place   
11     Roosevelt Island                  Park           Coffee Shop   
13       Lincoln Square                  Café          Concert Hall   
26  Morningside Heights                  Café             Bookstore   
28    Battery Park City                  Park                 Hotel   
35           Turtle Bay    Italian Restaurant                  Park   
36           Tudor City                  Park                  Café   
39         Hudson Yards   American Restaurant  Gym / Fitness Center   

   3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
2                 Bakery    Chinese Restaurant      Mobile Phone Shop   
3                 Lounge                  Bank            Pizza Place   
5     Mexican Restaurant    Italian Restaurant             Food Court   
6   Gym / Fitness Center     French Restaurant                    Bar   
11     Food & Drink Shop        Scenic Lookout            Supermarket   
13                 Plaza               Theater  Performing Arts Venue   
26        Sandwich Place   American Restaurant          Deli / Bodega   
28           Coffee Shop                   Gym          Memorial Site   
35         Deli / Bodega            Steakhouse            Coffee Shop   
36    Mexican Restaurant         Deli / Bodega                   Bank   
39           Coffee Shop                  Park          Boat or Ferry   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
2          Grocery Store                  Bank                  Park   
3                   Park         Deli / Bodega                Bakery   
5       Cuban Restaurant        Shipping Store            Bike Trail   
6     Seafood Restaurant   Fried Chicken Joint   American Restaurant   
11                  Café              Bus Line                   Gym   
13  Gym / Fitness Center                Bakery   Indie Movie Theater   
26                  Park           Coffee Shop          Burger Joint   
28            Food Court                 Plaza         Boat or Ferry   
35      Sushi Restaurant    Seafood Restaurant     French Restaurant   
36  Gym / Fitness Center      Sushi Restaurant       Thai Restaurant   
39                 Hotel                   Gym    Italian Restaurant   

      9th Most Common Venue 10th Most Common Venue  
2   New American Restaurant        Supplement Shop  
3                  Wine Bar   Caribbean Restaurant  
5                     Plaza             Food Truck  
6                      Bank                Library  
11     Gym / Fitness Center        Bubble Tea Shop  
13                     Park      French Restaurant  
26              Gaming Cafe      Outdoor Sculpture  
28               Playground            Pizza Place  
35           Sandwich Place     Turkish Restaurant  
36                  Dog Run             Restaurant  
39             Burger Joint             Restaurant

#### Cluster 4

In [61]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 3, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
37  Stuyvesant Town         Boat or Ferry                  Park   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
37           Yoga Studio        Farmers Market                   Bar   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
37              Fountain           Coffee Shop          Cocktail Bar   

   9th Most Common Venue 10th Most Common Venue  
37           Gas Station                 Bistro

#### Cluster 5

In [62]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 4, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0  Marble Hill        Sandwich Place      Department Store   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0       Supplement Shop      Storage Facility            Steakhouse   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0        Clothing Store           Coffee Shop    Seafood Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0         Deli / Bodega            Yoga Studio

# Results and Discussion

The neighborhoods of Toronto are very multicultural. There are a lot of different cuisines including Middle Eastern, Moroccan, Latin American, Italian, Japanese, Thai and Chinese. Toronto	 seems to take a step further in this direction by having a lot of restaurants, pub, cafe, coffee shops, sandwich place, breakfast spots, ice cream shop, frozen yogurt shop etc. For leisure, the neighborhoods are set up to have gym, skating rink, lounge, yoga studio, dance studio, playground and monument/landmark. Overall, the city of Toronto offers a multicultural, diverse and certainly an entertaining experience.

New York City has a wide variety of cuisines and eateries including French, African, Japanese, Korean, Chinese etc. There are a lot of hangout spots including café, coffee shop, wine shop and cocktail bars. It has a lot of shopping options too with that of the shopping mall, cosmetics shop, bakery, optical shop etc. For leisure and sight-seeing, there are a lot of parks, playground, yoga studio, gym/fitness center, spa, memorial sites and art gallery. Overall, New York City seems like the relaxing vacation spot with a mix of parks, historic spots and a wide variety of cuisines to try out.

# Conclusion

The purpose of this project was to explore the cities of Toronto and New York City and see how attractive it is to potential tourists and migrants. We explored both the cities based on their postal codes and then extrapolated the common venues present in each of the neighborhoods. Finally, we conclude with clustering similar neighborhoods together.

We could see that each of the neighborhoods in both the cities have a wide variety of experiences to offer which is unique in its own way. The cultural diversity is the evident which also gives the feeling of a sense of inclusion.
Both Toronto and New York City seem to offer a vacation stay or a romantic gateway with a lot of places to explore, beautiful landscapes and a wide variety of culture. In a nutshell, it depends on the stakeholders to decide which experience they would prefer more, and which would more to their liking.
